### DISCUSSION TITLE
About character width

### DISCUSSION THREAD

**Discussion by q962** [ryanoasis/nerd-fonts#969] [Q&A] [answered]
I'm using `JetBrains Mono Medium Nerd Font Complete.ttf` in GTK and found that there is a problem with the character width.
![image](https://user-images.githubusercontent.com/15686075/195878275-be40ce91-4371-4d8e-85ba-ad7b6b64473e.png)

![image](https://user-images.githubusercontent.com/15686075/195880291-598cf649-6aed-4b99-8daf-843a8a5a6b46.png)

Use fontforge to view the character 'F661':
![image](https://user-images.githubusercontent.com/15686075/195879630-7592a256-7a32-4f62-9c74-1fc4391a47cb.png)

By comparing the width of the characters displayed in gtk with the width of the characters displayed in fontforge, is it possible to determine that the problem is with the font?

![image](https://user-images.githubusercontent.com/15686075/195881534-a240d601-ddb6-434d-ac35-a8a8ff3ab273.png)

I looked at 'JetBrains Mono Medium Nerd Font Complete Mono.ttf' again, and although its character width is still 600, the characters are within the set character width.

Version 2.2.2

**Comment by Finii** [comment, marked as answer]
This is a good question to summarize all the facts (it is spread all over at the moment I think).

_(Please ignore all line spacings in my example images, assume they are correct. I was just too laze to create 'correct' images and it has not really any relevance to what I talk about here. It is always about left-right dimensions a.k.a. width.)_

The aim [1] of Nerd Fonts is to provide fonts for programming. Most programming happens with monospaced fonts. That means that all letters have the same width and what you in effect get is a 'grid' of characters. Like this

![image](https://user-images.githubusercontent.com/16012374/195900568-7a301b8b-240a-492c-aff1-343683ea384a.png)

Monospaced: Each letter (glyph) has the same width
Proportional: Each glyph has the room it needs, small are small and wide have more room

But now we want to have some symbols. Like the github logo

![image](https://user-images.githubusercontent.com/16012374/195901641-3fbda6ec-f7f1-4385-a0aa-7ae0b6d7467a.png)

If we insert that into a monospaced font and keep the monospaced-promise (all stuff has the same width), we have this:

![image](https://user-images.githubusercontent.com/16012374/195902986-e531357a-c969-487c-99ad-660b7c68085e.png)

There are a lot of circular icons like that. Usually letters are taller than wide. That means for monospaced fonts that the icon needs to be scaled waaay down to fit the width. It appears rather small and not tall enough. If the icon would be scaled for correct tallness (like in the proportional example) it would fill more than one width: The font ist not monospaced.

So, now we identified one way to create a patched font: The strictly monospaced font (on the right), which is called `Nerd Font Mono` (_JetBrains Mono Medium **Nerd Fonts** Complete **Mono**_)

"But the icons are so small!"

When we make the icons bigger we have two possibilities:

![image](https://user-images.githubusercontent.com/16012374/195904679-a05cf3b9-3657-4fae-a028-7eb913261592.png)

We can insert the icon with its correct width, like on the left side. The icon is just wider than normal letters: We do not need to take any precautions, but ... our strict grid is gone. Furthermore a lot of programming text editors would not take that font. They want monospaced fonts and you can not even select that font, or the oversized glyphs are scaled down by the client to fit the 'grid'!
The left font would be a `Nerd Font Proportional` (which is not existing jet, see links below).

"But I want to use my usual terminal emulator with the fonts!"

That is where the right solution comes in. Here the icon's width is 'standard width', but when the icon is drawn it extends into the next space. You show one example as topmost image, the cloud that seems to be bigger than one 'cell':

![image](https://user-images.githubusercontent.com/16012374/195906089-e368dff4-cb2a-44cd-a56b-605753b4acec.png)

Now all glyphs in the font have the same width and the font is detected as 'monospaced'. It can be used in all the places. But ... the user has to make sure that there is enough free room after the icon. This method has become so widespread that a lot of programs assume this. They insert a 'blank' after a symbol. Terminal emulators detect them as '2 cells wide' and handles them accordingly. Sometime they decide on actual blanks after the icon. This is the ordinary `Nerd Font`. Bigger symbols but still fully monospaced.

If you use the font in a GTK context, the best would be a `Nerd Font Proportional` for you. Or even better, you should insert the symbol not from your text font but from the special `Nerd Font Symbols Only` font. That font is designed to be used together with other (unpatched) text fonts. So in your GTK application just switch font for the icons.

Here are some topics that also talk about this. The Nerd-Font-naming will be different in some.

* https://github.com/ryanoasis/nerd-fonts/issues/746#issuecomment-1003694798
* #731 
* https://github.com/ryanoasis/nerd-fonts/issues/520#issuecomment-1007317631
* https://github.com/ryanoasis/nerd-fonts/issues/898#issuecomment-1230128066
* https://github.com/ryanoasis/nerd-fonts/discussions/951#discussioncomment-3781642
* [`font-patcher: Add possibility to generate proportional font`](https://github.com/ryanoasis/nerd-fonts/pull/764/commits/63db683b1a6c4a7341ec5872f437a8953a321bb0) 63db683b1
* #900


There a lot more Issues, all wrought around this problems, and it usually turns out as 'terminal emulator user' vs. 'GUI user'.

[1] Well, that does not hold anymore I believe. But let's call it "one of the prime goals" or such.

_Edit: Add footnote_

**Reply by q962** [reply]
Interesting design.
I don't know anything about the purpose of this design. Never seen a related article somewhere. 
I know very little about every aspect of fonts, even naming conventions.
Thanks for answering.

**Comment by Finii** [comment]
Remark: Some might think that the `Nerd Font` solution (having the glyph extend out of the 'cell') is strange and kind of cheating. Well it is not and a widespread font feature, at least in script fonts. See some example in https://github.com/ryanoasis/nerd-fonts/pull/748#issuecomment-1240407375

![image](https://user-images.githubusercontent.com/16012374/196033566-cf2371c5-799b-41d1-90ae-d514b4016175.png)

Here, also, it is the author's duty to check that there is no ugly overlap in the output.
